packages/ai/src/ai/common/sandbox.py

```
@@ -0,0 +1,252 @@
+    }
+
+    # ── 2. Inject allowlist-only __import__ ──────────────────────────────
+    original_import = builtins.__import__
+
+    def restricted_import(name: str, *args: Any, **kwargs: Any) -> Any:
+        top_level = name.split('.')[0]
+        if top_level not in allowlist:
+            raise ImportError(
+                f"Import of '{name}' is not allowed. "
+                f"Allowed modules: {', '.join(sorted(allowlist))}"
+            )
```

Ruff (0.15.5)

[warning] Avoid specifying long messages outside the exception class

(TRY003)